src/panels/lovelace/editor/card-editor/hui-card-editor.ts

```diff
@@ -0,0 +1,264 @@
+  @property() private _error?: string;
+  // Warning: GUI editor can't handle configuration - ok to save
+  @property() private _warning?: string;
+  @property() private _loading: boolean = false;
+
+  public get yaml(): string {
+    return this._yaml || "";
+  }
+  public set yaml(_yaml: string) {
+    this._yaml = _yaml;
+    this._updateConfigElement();
+    // TODO: Defer this to next redraw? How?
+    setTimeout(() => this._yamlEditor.codemirror.refresh(), 1);
+    fireEvent(this, "config-changed", { config: this.config! });
+  }
+
+  public get config(): LovelaceCardConfig | undefined {
+    try {
+      return yaml.safeLoad(this.yaml);
+    } catch (err) {
+      this._error = err;
+      return { type: "error", error: err.message };
+    }
+  }
+  public set config(config: LovelaceCardConfig | undefined) {
+    this.yaml = yaml.safeDump(config);
+  }
+
+  public get error(): boolean {
+    return this._error !== undefined;
+  }
+
+  private get _yamlEditor(): HuiYamlEditor {
+    return this.shadowRoot!.querySelector("hui-yaml-editor")!;
+  }
+
+  protected render(): TemplateResult {
+    return html`
+      <div style="width: 100%;">
+        <div class="gui-editor" ?hidden="${this._GUImode === false}">
+          ${this._loading
+            ? html`
+                <paper-spinner
+                  ?active="${this._loading}"
+                  alt="Loading"
+                  class="center margin-bot"
+                ></paper-spinner>
+              `
+            : this._warning || this._error
+            ? ""
+            : this._configElement}
+        </div>
+        <div class="yaml-editor" ?hidden="${this._GUImode === true}">
+          <hui-yaml-editor
+            .hass="${this.hass}"
+            .value="${this.yaml}"
+            @yaml-changed="${this._handleYAMLChanged}"
+          ></hui-yaml-editor>
+          <div class="error">
+            ${this._error}
+          </div>
+          <div class="warning">
+            ${this._warning}
+          </div>
+        </div>
+        <div class="buttons">
+          <mwc-button
+            @click="${() => (this._GUImode = !this._GUImode)}"
+            ?disabled="${this._warning || this._error}"
+            ?unelevated="${this._GUImode === false}"
+          >
+            <ha-icon .icon="${"mdi:code-braces"}"></ha-icon>
+          </mwc-button>
+        </div>
+      </div>
+    `;
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      :host {
+        margin: 0 10px;
+        display: flex;
+      }
+      .hidden {
+        display: none;
+      }
+      .gui-editor,
+      .yaml-editor {
+        padding: 8px 0px;
+      }
+      .error {
+        color: #ef5350;
+      }
+      .warning {
+        color: #ffa726;
+      }
+      .buttons {
+        text-align: right;
+
+        padding: 8px 0px;
+        border-top: 1px solid #e8e8e8;
+      }
+    `;
+  }
+
+  protected async firstUpdated() {
+    this.config = this.value;
```

Comment:
Is this necessary? Shouldn't we define a getter that is `this.config || this.value` ?

Reply:
Turns out the `value` property can be removed and `config` used in the interface.
That didn't work before because the editor was embedded in an element that kept updating and overwriting the changes.

Reply:
Nope. I was wrong. This is needed.

`value` needs to be transfered to `config` only once. Otherwise it will be overwritten by the embedding element when an error is returned (from `get config()`).

`this.config || this.value` won't work, because `this.config` just points to `this.yaml` and doing this for that would get weird in the off chance that a user opens the yaml editor and clears it.

Reply:
This does add the side effect that it is no longer possible to externally update the value of this element.

Reply:
I think that we should move this to `updated` and do it each time `value` has changed